File: tools/ad-preview.js
#!/usr/bin/env node
/* The simulated paid bands, shot inside the real pages.
 *
 *     python3 tools/ad-preview.py && node tools/ad-preview.js
 *
 * NOT A GATE. It writes images and asserts nothing, because the question it
 * exists for is whether a sponsored band reads as bought rather than as
 * editorial, and no count can answer that. The pages come from
 * `.cache/ad-preview` and everything else — the stylesheet, the maps, the
 * photographs — is served from `site/`, so what is photographed is the real
 * page with one band added and nothing else changed.
 *
 * It refuses a page that is not 200 and a page with no band, for the reason
 * the contact sheet refuses a 404: a blank cell still looks like a result.
 */
const { chromium } = require("playwright");
const fs = require("fs");
const http = require("http");
const path = require("path");

const ROOT = path.resolve(__dirname, "..");
const ROOTS = [path.join(ROOT, ".cache", "ad-preview"), path.join(ROOT, "site")];
const EXE = process.env.HERO_SHEET_EXE || "/opt/pw-browsers/chromium";
const TYPES = {
  ".html": "text/html; charset=utf-8", ".css": "text/css", ".js": "text/javascript",
  ".json": "application/json", ".jpg": "image/jpeg", ".png": "image/png",
  ".webp": "image/webp", ".avif": "image/avif", ".svg": "image/svg+xml",
};
const PAGES = [["home", "/"], ["bergen", "/europe/norway/fjord-norway/bergen"]];

(async () => {
  const server = http.createServer((req, res) => {
    const url = decodeURIComponent(req.url.split("?")[0]);
    for (const base of ROOTS) {
      for (const c of [path.join(base, url), path.join(base, url, "index.html")]) {
        if (fs.existsSync(c) && fs.statSync(c).isFile()) {
          res.writeHead(200, { "Content-Type": TYPES[path.extname(c)] || "application/octet-stream" });
          return res.end(fs.readFileSync(c));
        }
      }
    }
    res.writeHead(404, { "Content-Type": "text/plain" });
    res.end("not here");
  });
  await new Promise((r) => server.listen(0, "127.0.0.1", r));
  const BASE = `http://127.0.0.1:${server.address().port}`;
  const browser = await chromium.launch({ executablePath: EXE });
  const out = [];
  for (const [name, url] of PAGES) {
    for (const [suffix, viewport] of [
      ["wide", { width: 1280, height: 900 }],
      ["phone", { width: 390, height: 844 }],
    ]) {
      const page = await browser.newPage({ viewport });
      const r = await page.goto(BASE + url, { waitUntil: "networkidle" });
      if (r.status() !== 200) throw new Error(`${url} answered ${r.status()}`);
      const band = page.locator(".adband").first();
      if (!(await band.count())) throw new Error(`${url} drew no paid band`);
      /* LAZY IMAGES DO NOT LOAD FOR A SCREENSHOT TAKEN WITHOUT SCROLLING,
       * which this repository already records: `loading="lazy"` keys on the
       * viewport, so a band below the fold photographs as empty boxes and an
       * empty box under a scrim looks exactly like a missing photograph. */
      await band.scrollIntoViewIfNeeded();
      await page.waitForTimeout(400);
      const file = `ad-preview-${name}-${suffix}.png`;
      /* The band AND what is above it, because the question is whether a
       * reader can tell the two apart — a photograph of the band alone
       * answers "is it legible" and not "is it distinguishable". */
      await page.screenshot({ path: file, clip: await (async () => {
        const b = await band.boundingBox();
        const top = Math.max(0, b.y - 360);
        return { x: 0, y: top, width: viewport.width, height: Math.min(b.y + b.height + 40 - top, 1400) };
      })() });
      out.push(file);
      await page.close();
    }
  }
  await browser.close();
  server.close();
  console.log(out.join("\n"));
})();

File: tools/ad-preview.py
#!/usr/bin/env python3
"""What a paid band looks like, on the real pages, with nothing acquired.

    python3 tools/ad-preview.py && node tools/ad-preview.js

NOT A GATE, AND FOR THE SAME REASON THE HERO SHEET IS NOT ONE: it writes an
image rather than a verdict, and the step it exists for is a person looking.
Every gate here can say the commercial layer is off and correct; none of them
can say whether a sponsored band reads as bought rather than as editorial,
and that is the whole design question §17 and §33 are about.

It writes into `.cache/`, which is ignored, for the reason the candidate
previews do: this is not the site, nothing here is served, and a page
carrying an advertisement must never be mistaken for something the build
produced. `site/` is not touched and `data/advertising.json` is not written.
"""

import os
import shutil
import sys

sys.path.insert(0, os.path.dirname(os.path.abspath(__file__)))

from lib import addemo                 # noqa: E402
from lib import data as D              # noqa: E402
from lib import pages as P             # noqa: E402

ROOT = os.path.dirname(os.path.dirname(os.path.abspath(__file__)))
OUT = os.path.join(ROOT, ".cache", "ad-preview")


def main():
    d = D.load()
    shutil.rmtree(OUT, ignore_errors=True)
    with addemo.simulated():
        wrote = []
        path, html = P.home(d)
        wrote.append((path, html))
        # Bergen, because it is the destination every instrument here uses
        # as the shape of that family and the one the accessibility scan
        # already represents it with.
        c = d["countries"]["norway"]
        r = next(x for x in c["regions"] if x["slug"] == "fjord-norway")
        t = next(x for x in r["cities"] if x["slug"] == "bergen")
        path, html = P.city_page(d, c, r, t)
        wrote.append((path, html))
    for path, html in wrote:
        dest = os.path.join(OUT, path.lstrip("/"))
        os.makedirs(os.path.dirname(dest), exist_ok=True)
        with open(dest, "w", encoding="utf-8") as f:
            f.write(html)
        band = html.count('class="adband"')
        print(f"  {path}  {len(html):,} bytes, {band} paid band(s)")
        if band != 1:
            raise SystemExit(
                f"{path} rendered {band} bands: the preview exists to show "
                f"one, and a page with none is the OFF state photographed "
                f"by accident")
    print(f"\n{OUT} — now: node tools/ad-preview.js")


if __name__ == "__main__":
    main()
